src/hooks/resources/device.ts

```diff
@@ -141,25 +140,51 @@ sbvrUtils.addPureHook('POST', 'resin', 'device', {
 		// Keep the app ID for later -- we'll need it in the POSTRUN hook
 		request.custom.appId = request.values.belongs_to__application;
 
-		// transform to canonical slug in case the UI and API are out of sync
-		waitPromises.push(
-			deviceTypes
-				.normalizeDeviceType(request.values.device_type)
-				.then(deviceType => {
-					request.values.device_type = deviceType;
-					request.values.device_name =
-						request.values.device_name || haikuName.generate();
-					request.values.uuid =
-						request.values.uuid || crypto.pseudoRandomBytes(31).toString('hex');
-
-					if (!/^[a-f0-9]{32}([a-f0-9]{30})?$/.test(request.values.uuid)) {
-						throw new BadRequestError(
-							'Device UUID must be a 32 or 62 character long lower case hex string.',
+		request.values.device_name =
+			request.values.device_name || haikuName.generate();
+		request.values.uuid =
+			request.values.uuid || crypto.pseudoRandomBytes(31).toString('hex');
+
+		if (!/^[a-f0-9]{32}([a-f0-9]{30})?$/.test(request.values.uuid)) {
+			throw new BadRequestError(
+				'Device UUID must be a 32 or 62 character long lower case hex string.',
+			);
+		}
+
+		return Promise.try(() => {
+			if (
+				request.values.device_type != null &&
+				request.values.is_of__device_type == null
+			) {
+				// translate device_type to is_for__device_type
+				return deviceTypes
+					.getDeviceTypeIdBySlug(api, request.values.device_type)
+					.then(dt => {
+						if (dt) {
+							request.values.is_of__device_type = dt.id;
+							return;
+						}
+						throw new deviceTypes.UnknownDeviceTypeError(
+							request.values.device_type,
 						);
-					}
-				}),
-		);
-		return Promise.all(waitPromises);
+					});
+			}
+		})
+			.then(() => {
+				return api.get({
+					resource: 'device_type',
+					id: request.values.is_of__device_type,
+					options: {
+						$select: ['slug'],
+					},
+				});
+			})
+			.then((dt: { slug: string }) => {
+				if (!dt) {
+					throw new deviceTypes.InvalidDeviceTypeError();
+				}
+				return deviceTypes.findBySlug(api, dt.slug);
+			});
```

Comment:
All this added code seems to be identical in POST and PATCH hooks -- can you please factor it out in a shared function that's called from both places?

Reply:
I think you mean between the application and device resource? I refactored these to a shared functions.

Reply:
Yes, sorry. I saw the changes and they look really good!